L1Trigger/L1THGCal/plugins/geometries/HGCalTriggerGeometryV9Imp3.cc

```diff
@@ -0,0 +1,1105 @@
+    edm::LogError("HGCalTriggerGeometryV9Imp3")
+        << "The mapping input json file does not have the expected structure for the Module block";
+  }
+
+  json_input_file.close();
+}
+
+unsigned HGCalTriggerGeometryV9Imp3::packLayerSubdetWaferId(unsigned layer, int subdet, int waferU, int waferV) const {
+  unsigned packed_value = 0;
+
+  packed_value |= ((waferU & HGCalModuleDetId::kHGCalModuleUMask) << HGCalModuleDetId::kHGCalModuleUOffset);
+  packed_value |= ((waferV & HGCalModuleDetId::kHGCalModuleVMask) << HGCalModuleDetId::kHGCalModuleVOffset);
+  packed_value |= ((subdet & HGCalModuleDetId::kHGCalSubdetMask) << HGCalModuleDetId::kHGCalSubdetOffset);
+  packed_value |= ((layer & HGCalModuleDetId::kHGCalLayerMask) << HGCalModuleDetId::kHGCalLayerOffset);
+  return packed_value;
+}
+
+void HGCalTriggerGeometryV9Imp3::unpackLayerSubdetWaferId(
+    unsigned wafer, unsigned& layer, int& subdet, int& waferU, int& waferV) const {
+  waferU = (wafer >> HGCalModuleDetId::kHGCalModuleUOffset) & HGCalModuleDetId::kHGCalModuleUMask;
+  waferV = (wafer >> HGCalModuleDetId::kHGCalModuleVOffset) & HGCalModuleDetId::kHGCalModuleVMask;
+  subdet = (wafer >> HGCalModuleDetId::kHGCalSubdetOffset) & HGCalModuleDetId::kHGCalSubdetMask;
+  layer = (wafer >> HGCalModuleDetId::kHGCalLayerOffset) & HGCalModuleDetId::kHGCalLayerMask;
+}
+
+void HGCalTriggerGeometryV9Imp3::etaphiMappingFromSector0(int& ieta, int& iphi, unsigned sector) const {
+  if (sector == 0) {
+    return;
+  }
+  iphi = iphi + (sector * hSc_num_panels_per_sector_);
+}
+
+HGCalGeomRotation::WaferCentring HGCalTriggerGeometryV9Imp3::getWaferCentring(unsigned layer, int subdet) const {
+  if (subdet == ForwardSubdetector::HGCEE) {  // CE-E
+    return HGCalGeomRotation::WaferCentring::WaferCentred;
+  } else if (subdet == ForwardSubdetector::HGCHEF || subdet == ForwardSubdetector::HGCHEB) {
+    if ((layer % 2) == 1) {  // CE-H Odd
+      return HGCalGeomRotation::WaferCentring::CornerCentredY;
+    } else {  // CE-H Even
+      return HGCalGeomRotation::WaferCentring::CornerCentredMercedes;
+    }
+  } else {  //HFNose
+    return HGCalGeomRotation::WaferCentring::WaferCentred;
+  }
+}
+
+unsigned HGCalTriggerGeometryV9Imp3::etaphiMappingToSector0(int& ieta, int& iphi) const {
+  unsigned sector = 0;
+
+  if (unsigned(std::abs(iphi)) > 2 * hSc_num_panels_per_sector_)
+    sector = 2;
+  else if (unsigned(std::abs(iphi)) > hSc_num_panels_per_sector_)
+    sector = 1;
+  else
+    sector = 0;
+
+  iphi = iphi - (sector * hSc_num_panels_per_sector_);
+
+  return sector;
+}
+
+bool HGCalTriggerGeometryV9Imp3::validTriggerCell(const unsigned trigger_cell_id) const {
+  return validTriggerCellFromCells(trigger_cell_id);
+}
+
+bool HGCalTriggerGeometryV9Imp3::disconnectedModule(const unsigned module_id) const {
+  bool disconnected = false;
+  if (disconnected_layers_.find(layerWithOffset(module_id)) != disconnected_layers_.end())
+    disconnected = true;
+  return disconnected;
```

Comment:
It will be necessary to add a check for modules not linked to stage-1 FPGAs. Such modules should be considered as disconnected.

Reply:
thanks - check added